### PR TITLE
feat: epoch analysis tool

### PR DESCRIPTION
**Pull request by Longarithm**
Tool to analyse epoch infos in two modes:
* `check-consistency` - regenerate next next epoch info based on two previous epochs and check that it matches the epoch info stored in DB;
* `backtest` - regenerate epoch infos with existing proposals, rewards and kickouts as if `PROTOCOL_VERSION` was always in place.

The `backtest` was used to estimate new algorithm for chunk producer shard assignments and showed that on average there is only one state sync happening, if we use start epoch height >= 545. Epoch info for 544 can't be retrieved for some reason, see #11477. 

Consistency check revealed that some epochs in the past can't be replayed, see #11476.
Expected output of it, note that epoch T generates epoch T+2:
```
$ neard view-state epoch-analysis --start-height 1359 check-consistency
HEIGHT | VERSION | STATE SYNCS
  1361 |      53 |           0
  1362 |      53 |           8
  1363 |      54 |           8
  1364 |      54 |           8
  1365 |      54 |          24
  1366 |      54 |          12
...
```

### CODE REVIEW

File: tools/state-viewer/src/cli.rs [hunk 1]
```diff
@@ -486,6 +491,19 @@ impl EpochInfoCmd {
     }
 }
 
+#[derive(clap::Args)]
+pub struct EpochAnalysisCmd {
+    /// The height of the epoch to analyze.
+    #[clap(long)]
+    height: EpochHeight,
```
Comment (tayfunelmas) [inline tools/state-viewer/src/cli.rs:498]: should this be the start height of the epoch or any height in that epoch?

**Comment by tayfunelmas** [comment]
> WIP. cc @tayfunelmas if you are curious how it should work :)

Thanks for sharing, I was indeed curious about this part (how to iterate over the epoch-related info)

File: tools/state-viewer/src/cli.rs (hunk 1 above)
Comment (Longarithm) [reply tools/state-viewer/src/cli.rs]: Start height of the epoch.

File: tools/state-viewer/src/commands.rs [hunk 2]
```diff
@@ -896,6 +898,207 @@ pub(crate) fn print_epoch_info(
...
+    for (epoch_height, epoch_info) in
+        epoch_heights_to_infos.range(min_epoch_height..=max_epoch_height - 4)
+    {
+        let next_epoch_height = epoch_height.saturating_add(1);
+        let next_next_epoch_height = epoch_height.saturating_add(2);
+        let next_epoch_id = epoch_heights_to_ids.get(&next_epoch_height).unwrap();
+        let next_next_epoch_id = epoch_heights_to_ids.get(&next_next_epoch_height).unwrap();
+        let epoch_summary = epoch_heights_to_validator_infos.get(epoch_height).unwrap();
+
+        next_epoch_info = match mode {
```
Comment (wacban) [inline tools/state-viewer/src/commands.rs:977]: That's a fun way to implement where you have multiple of those checks :) I'm totally fine with it but I'm curious how did you arrive at this and how does it compare to having two methods, one for each mode?

**Review by wacban — COMMENTED** [review]
LGTM but I I want to understand the early return in `get_initial_chunk_producer_assignment` before approving - see the comment please.

File: chain/epoch-manager/src/shard_assignment.rs [hunk 3]
```diff
@@ -128,13 +128,20 @@ fn get_initial_chunk_producer_assignment(
...
             let chunk_producer_index = chunk_producer_indices.get(validator_stake.account_id());
-            if let Some(&index) = chunk_producer_index {
-                chunk_producers.push(index);
+            let Some(&index) = chunk_producer_index else {
+                continue;
+            };
+
+            chunk_producers.push(index);
+            if used_chunk_producers.contains(&index) {
+                return vec![vec![]; num_shards as usize];
```
[the remark this answers is not in the document]
Comment (Longarithm) [reply chain/epoch-manager/src/shard_assignment.rs]: Actually it was a leftover of implementation with some bug, removed.

File: tools/state-viewer/src/commands.rs (hunk 2 above)
Comment (Longarithm) [reply tools/state-viewer/src/commands.rs]: That was just bad code, how about now?
These two modes really share a lot of code. The difference is - for the first one, I always take data from DB. For the second, I override some data with previously generated info to simulate new algorithm behaviour. But the rest stays the same, so I don't want to separate impls completely.

File: tools/state-viewer/src/commands.rs [hunk 4]
```diff
@@ -896,6 +898,207 @@ pub(crate) fn print_epoch_info(
...
+            .map(|(epoch_id, epoch_info)| (epoch_info.epoch_height(), epoch_id.clone())),
+    );
+    let epoch_heights_to_infos =
+        BTreeMap::from_iter(epoch_infos.values().map(|e| (e.epoch_height(), e.clone())));
+    let min_epoch_height = epoch_height;
+    let max_epoch_height = *epoch_heights_to_infos.keys().max().unwrap();
+    let epoch_heights_to_validator_infos =
+        BTreeMap::from_iter(epoch_heights_to_ids.iter().filter_map(|(epoch_height, epoch_id)| {
+            if *epoch_height >= min_epoch_height
+                && *epoch_height <= max_epoch_height.saturating_sub(4)
```
[the remark this answers is not in the document]
Comment (Longarithm) [reply tools/state-viewer/src/commands.rs]: Good point, I can actually put 2 there, made it more clear.

**Review by wacban — APPROVED** [review]
LGTM

**Pull request merged**